File: trunk/Noxa.Emulation.Psp.Bios.NativeHLE/Kernel_Scheduling.cpp
// ----------------------------------------------------------------------------
// PSP Player Emulation Suite
// Copyright (C) 2006 Ben Vanik (noxa)
// Licensed under the LGPL - see License.txt in the project root for details
// ----------------------------------------------------------------------------

#include "Stdafx.h"
#define WIN32_LEAN_AND_MEAN
#include <windows.h>
#include "Kernel.h"
#include "NativeBios.h"
#include "HandleTable.h"
#include "KThread.h"
#include "KEvent.h"
#include "KCallback.h"
#include "Module.h"

using namespace Noxa::Emulation::Psp;
using namespace Noxa::Emulation::Psp::Bios;
using namespace Noxa::Emulation::Psp::Cpu::Native;

bool Kernel::Schedule()
{
	while( this->SchedulableThreads->GetCount() == 0 )
	{
		// No threads to run? Check for delayed threads
		void* state = NULL;
		KThread** pthread;
		KThread* earliest = NULL;
		while( ( pthread = Threads->Enumerate( &state ) ) != NULL )
		{
			KThread* thread = *pthread;
			if( thread->State == KThreadWaiting )
			{
				if( thread->WaitingOn == KThreadWaitDelay )
				{
					if( ( earliest == NULL ) ||
						( thread->WaitTimestamp + thread->WaitTimeout ) > ( earliest->WaitTimestamp + thread->WaitTimeout ) )
						earliest = thread;
				}
			}
		}

		if( earliest != NULL )
		{
			// Wait on it
			while( earliest->State == KThreadWaiting )
				Sleep( 1 );
		}
		else
		{
			Debug::WriteLine( "Kernel::Schedule: ran out of threads to run - exiting sim" );
			return false;
		}
	}

	// Find the next thread to run - this is easy
	KThread* nextThread = this->SchedulableThreads->PeekHead();
	assert( nextThread != NULL );

	// If it didn't change, don't do anything
	if( nextThread == ActiveThread )
		return false;

	// Switch
	Cpu->SwitchContext( nextThread->Context->ContextID );

	ActiveThread = nextThread;

	return true;
}

void Kernel::Execute()
{
	if( ActiveThread != NULL )
	{
		// Execute active thread
		int breakFlags;
		int instructionCount = Cpu->Execute( &breakFlags );
		if( breakFlags != 0 )
		{
			// Schedule request
			this->Schedule();
		}

		//Debug::WriteLine( String::Format( "Kernel: execute ended after {0} instructions w/ flags {1}", instructionCount, breakFlags ) );
	}
	else
	{
		// Load game and run until it creates its first thread

		IEmulationInstance^ emu = Emu;
		Debug::Assert( emu != nullptr );
		NativeBios^ bios = Bios;
		Debug::Assert( bios != nullptr );
		ILoader^ loader = bios->Loader;
		Debug::Assert( loader != nullptr );

		// Clear everything
		emu->LightReset();

		// Get bootstream
		Debug::Assert( bios->BootStream == nullptr );
		GameLoader^ gameLoader = gcnew GameLoader();
		bios->BootStream = gameLoader->FindBootStream( bios->Game );
		Debug::Assert( bios->BootStream != nullptr );

		LoadParameters^ params = gcnew LoadParameters();
		params->Path = bios->Game->Folder;
		LoadResults^ results = loader->LoadModule( ModuleType::Boot, bios->BootStream, params );
		
		Debug::Assert( results->Successful == true );
		if( results->Successful == false )
		{
			Debug::WriteLine( "Kernel: load failed!" );
			bios->Game = nullptr;
			return;
		}

		this->CurrentPath = bios->Game->Folder;
		emu->Cpu->SetupGame( bios->Game, bios->BootStream );

		// Start modules
		for( int n = 0; n < bios->_moduleList->Count; n++ )
			bios->_moduleList[ n ]->Start();

		//// Run until we have a thread
		//int preThreadCount = 0;
		//while( ActiveThread == NULL )
		//{
		//	int breakFlags;
		//	preThreadCount += Cpu->Execute( &breakFlags );
		//	assert( breakFlags == 0 );
		//}

		//Debug::WriteLine( String::Format( "Kernel: first thread created, executed {0} instructions", preThreadCount ) );
	}
}

extern Kernel* _currentKernel;
void IssueCallbackComplete()
{
	// Let the kernel know where we are
	assert( _currentKernel != NULL );
	_currentKernel->CallbackComplete();
}

int Kernel::IssueCallback( int callbackId, int argument )
{
	KCallback* cb = ( KCallback* )Handles->Lookup( callbackId );
	if( cb == NULL )
		return -1;

	// Not sure how to handle a CB to an active thread!
	assert( cb->Thread != ActiveThread );

	_oldThread = ActiveThread;
	ActiveThread = cb->Thread;

	Cpu->MarshalCallback( cb->Thread->Context->ContextID, cb->Address, argument, &IssueCallbackComplete );

	// The CPU, when complete with the callback, will set the return value (hopefully)
	// It will then call IssueCallbackComplete so we can switch the thread back in ourselves

	return 0;
}

void Kernel::CallbackComplete()
{
	assert( _oldThread != NULL );

	ActiveThread = _oldThread;
	_oldThread = NULL;
}

File: trunk/Noxa.Emulation.Psp.Bios.ManagedHLE/Kernel.Scheduling.cs
// ----------------------------------------------------------------------------
// PSP Player Emulation Suite
// Copyright (C) 2006 Ben Vanik (noxa)
// Licensed under the LGPL - see License.txt in the project root for details
// ----------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text;

using Noxa.Emulation.Psp;
using Noxa.Emulation.Psp.Bios;
using Noxa.Emulation.Psp.Cpu;
using Noxa.Emulation.Psp.Games;

namespace Noxa.Emulation.Psp.Bios.ManagedHLE
{
	partial class Kernel
	{
		public bool Schedule()
		{
			while( this.SchedulableThreads.Count == 0 )
			{
				// No threads to run? Check for delayed
				KThread earliest = null;
				for( int n = 0; n < this.Threads.Count; n++ )
				{
					KThread thread = this.Threads[ n ];
					if( ( ( thread.State == KThreadState.Waiting ) || ( thread.State == KThreadState.WaitSuspended ) ) &&
						( thread.WaitingOn == KThreadWait.Delay ) )
					{
						if( ( earliest == null ) ||
							( ( thread.WaitTimestamp + thread.WaitTimeout ) > ( earliest.WaitTimestamp + earliest.WaitTimeout ) ) )
							earliest = thread;
					}
				}

				if( earliest != null )
				{
					// Wait on it
					while( earliest.State == KThreadState.Waiting )
					{
						// This happens A LOT - it'd be best if we just spun, but by sleeping we save some time
						System.Threading.Thread.Sleep( 0 );
					}
				}
				else
				{
					//Log.WriteLine( Verbosity.Verbose, Feature.Bios, "Schedule: ran out of threads to run" );
					return false;
				}
			}

			// Find next thread to run
			KThread nextThread = this.SchedulableThreads.Head;
			Debug.Assert( nextThread != null );

			// If it didn't change, don't do anything
			if( nextThread == ActiveThread )
				return false;

			// Switch
			Cpu.SwitchContext( nextThread.ContextID );

			ActiveThread = nextThread;

			return true;
		}

		public void Execute()
		{
			Debug.Assert( ActiveThread != null );

			// Execute active thread
			bool breakFlag;
			uint instructionsExecuted;
			Cpu.Execute( out breakFlag, out instructionsExecuted );
			if( breakFlag == false )
			{
				// Only if not broken by choice
				//Log.WriteLine( Verbosity.Verbose, Feature.Bios, "CPU returned to us after {0} instructions", instructionsExecuted );
			}
		}

		private KThread _oldThread;
		private KCallback _runningCallback;
		private bool _runningUserCall;

		public bool IssueCallback( KCallback callback, uint argument )
		{
			Debug.Assert( callback != null );
			Debug.Assert( _runningUserCall == false );
			Debug.Assert( _runningCallback == null );

			_oldThread = ActiveThread;
			ActiveThread = callback.Thread;

			_runningCallback = callback;

			// Format is arg1, arg2, commonAddress
			// arg1 = ?
			// arg2 is passed during notify and given to use in argument
			Cpu.MarshalCall( ActiveThread.ContextID, callback.Address, new uint[] { callback.NotifyCount, argument, callback.CommonAddress }, new MarshalCompleteDelegate( this.CallbackComplete ), 0 );

			return true;
		}

		private bool CallbackComplete( int tcsId, int state, int result )
		{
			Debug.Assert( _oldThread != null );
			Debug.Assert( _runningCallback != null );

			// Something ? return value something?

			ActiveThread = _oldThread;
			_oldThread = null;
			_runningCallback = null;

			return true;
		}

		public bool IssueUserCall( uint address, uint[] arguments )
		{
			Debug.Assert( _runningUserCall == false );
			Debug.Assert( _runningCallback == null );

			_runningUserCall = true;

			Cpu.MarshalCall( ActiveThread.ContextID, address, arguments, new MarshalCompleteDelegate( this.UserCallComplete ), ( int )address );

			return true;
		}

		private bool UserCallComplete( int tcsId, int state, int result )
		{
			Debug.Assert( _runningUserCall == true );

			uint address = ( uint )state;
			// Something ? return value?

			_runningUserCall = false;

			return true;
		}
	}
}
